spec/src/main/asciidoc/property-expression.adoc

```diff
@@ -0,0 +1,76 @@
+= Property Expression
+
+The property expression is in the format of ${a.property}. When resoving the property expression, the value of `a.property` will be used.
+With the property expression, it enables a property refer to another property.
+
+== An example
+
+Consider the following example:
+
+[source,properties]
+----
+server.url=http://${server.host}:${server.port}/hello
+server.host=my.org
+server.port=9080
+----
+
+When looking up `server.url`, the property value will be resoved to `http://my.org:9080/hello`.
```

Comment:
I assume that the property resolution must be performed relative to the root of the configuration source in which the property was found.

The reason for this assumption is that we did not settle on a definition of an "ultimate root" for configuration (that I recall - if we did, someone please correct me). However every configuration source must be somehow loaded in to a configuration which represents the root from the perspective of *that configuration source*; thus this makes the most sense to me as a resolution root.

Reply:
We discussed this on today's call. Mark, Tomas and I feel it is better to do the resolution globally so that some variables can be overridden in a different config source.
Tomas: If a property is used by multiple config sources, if this resolution only works locally, you have to repeat the definition of the same property in all config sources. Sometimes they might be out of sync, which causes more issues.

Reply:
When starting a `${` and ending with a `}`, it could be evaluated as an EL expression globally? eg.

server.url=${env.SERVER_URL}
server.port=${bean.ports[0]} //the first item of a CDI bean ports array/list